lsp-types/src/Language/LSP/Types/WorkspaceEdit.hs

```diff
@@ -38,12 +40,109 @@ deriveJSON lspOptions ''TextDocumentEdit
 
 -- ---------------------------------------------------------------------
 
+-- | For tagging `CreateFile`/`RenameFile`/`DeleteFile`
+-- Should this be merged with `ResourceOperationKind` ?
+data FileResourceChangeKind
+  = FileResourceChangeCreate
+  | FileResourceChangeRename
+  | FileResourceChangeDelete 
+  deriving (Read, Show, Eq)
+  
+instance ToJSON FileResourceChangeKind where
+  toJSON FileResourceChangeCreate = String "create"
+  toJSON FileResourceChangeRename = String "rename"
+  toJSON FileResourceChangeDelete = String "delete"
+
+instance FromJSON FileResourceChangeKind where
+  parseJSON (String "create") = pure FileResourceChangeCreate
+  parseJSON (String "rename") = pure FileResourceChangeRename
+  parseJSON (String "delete") = pure FileResourceChangeDelete
+  parseJSON _                 = mempty
```

Comment:
Bit of a nitpick here, but I think we can avoid having to create this data type if we omit storing the `kind` in `{Create,Rename,Delete}File`. Instead we can just parse it and emit the `kind` field in custom `ToJSON` instances, like how `WorkDoneProgressBegin` is handled

https://github.com/alanz/lsp/blob/8d6e8f316a5a1c40986cdc6400611376c8ab0140/lsp-types/src/Language/LSP/Types/Progress.hs#L89-L107

I think this more closely matches the string literal tags in TypeScript interfaces, and it means users can't accidentally create a dodgy `CreateFile FileResourceChangeRename x`

Reply:
Nice!